File: solutions/1997. First Day Where You Have Been in All the Rooms/1997.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  int firstDayBeenInAllRooms(const vector<int>& nextVisit) {
    constexpr int kMod = 1'000'000'007;
    const int n = nextVisit.size();
    vector<int> dp(n);

    for (int i = 1; i < n; ++i)
      dp[i] = (2L * dp[i - 1] - dp[nextVisit[i - 1]] + 2 + kMod) % kMod;

    return dp.back();
  }
};

File: solutions/1997. First Day Where You Have Been in All the Rooms/1997.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
  def firstDayBeenInAllRooms(self, nextVisit: List[int]) -> int:
    kMod = 1_000_000_007
    n = len(nextVisit)
    dp = [0] * n

    for i in range(1, n):
      dp[i] = (2 * dp[i - 1] - dp[nextVisit[i - 1]] + 2) % kMod

    return dp[-1]